### PR TITLE
osu! Difficulty Aim Overhaul: velocity change

### PR DESCRIPTION
**Pull request by Xexxar**
subset of https://github.com/ppy/osu/pull/14767 featuring focus 7

- [x] Depends on https://github.com/ppy/osu/pull/14845.
- [x] Depends on https://github.com/ppy/osu/pull/14846.

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs [hunk 1]
```diff
@@ -91,11 +96,31 @@ private double strainValueOf(DifficultyHitObject current)
...
 
-                    angleBonus = acuteAngleBonus * acute_angle_multiplier + wideAngleBonus * wide_angle_multiplier; // add the angle buffs together.
+                    angleBonus = Math.Max(acuteAngleBonus * acute_angle_multiplier, wideAngleBonus * wide_angle_multiplier); // Take the max of the multipliers.
                 }
             }
 
-            aimStrain += angleBonus; // Add in angle bonus.
+            if (Math.Max(prevVelocity, currVelocity) != 0)
+            {
+                prevVelocity = (osuLastObj.JumpDistance + osuLastObj.TravelDistance) / osuLastObj.StrainTime; // We want to use the average velocity when awarding differences, not necessarily combined.
```
Comment (apollo-dw) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:105]: comment would be better worded as `// We want to use the average velocity over the whole object when awarding differences, not the individual jump and slider path velocities.` since "not necessarily combined" is confusing considering they're being literally added together in the code

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs [hunk 2]
```diff
@@ -91,11 +96,31 @@ private double strainValueOf(DifficultyHitObject current)
...
+                currVelocity = (osuCurrObj.JumpDistance + osuCurrObj.TravelDistance) / osuCurrObj.StrainTime;
+
+                velChangeBonus = Math.Max(Math.Abs(prevVelocity - currVelocity) // reward for % distance slowed down compared to previous, paying attention to not award overlap
```
Comment (apollo-dw) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:108]: those whole math.max term should be split up into two variables, definitely hard to follow with how chunky this is

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs [hunk 3]
```diff
@@ -91,11 +96,31 @@ private double strainValueOf(DifficultyHitObject current)
...
+                                          * Math.Pow(Math.Sin(Math.PI / 2 * Math.Min(1, osuCurrObj.JumpDistance / 100)), 2) // do not award overlap
+                                          * Math.Pow(Math.Sin(Math.PI / 2 * Math.Abs(prevVelocity - currVelocity) / Math.Max(prevVelocity, currVelocity)), 2), // scale with ratio of difference compared to max
+                                          Math.Min(125 / Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime), Math.Abs(prevVelocity - currVelocity)) // reward for % distance up to 125 / strainTime for overlaps where velocity is still changing.
+                                          * Math.Pow(Math.Sin(Math.PI / 2 * Math.Abs(prevVelocity - currVelocity) / Math.Max(prevVelocity, currVelocity)), 2)); // scale with ratio of difference compared to max
```
Comment (apollo-dw) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:112]: take this term out of both maxes since it's present in both and multiply it afterwards

**Review by apollo-dw — CHANGES_REQUESTED** [review]
Seems to do as intended, just some readability stuff. I've recreated the whole thing (as it is right now) in desmos: https://www.desmos.com/calculator/cfdaruyy6x

Interested in seeing if anything crops up on Smoogi's spreadsheet, values I've seen so far with this change seem reasonable

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs (hunk 1 above)
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs:105]: done.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs (hunk 3 above)
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs]: done.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs (hunk 2 above)
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs]: done.

**Review by smoogipoo — APPROVED** [review]
Approved by @MBmasher / @apollo-dw
